lib/cloud/azure/imds.go

```diff
@@ -0,0 +1,192 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package azure
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/trace"
+)
+
+const (
+	// metadataReadLimit is the largest number of bytes that will be read from imds responses.
+	metadataReadLimit = 1_000_000
+	// imdsURL is the default base URL for Azure instance metadata.
+	imdsURL = "http://169.254.169.254/metadata"
+	// minimumSupportedAPIVersion is the minimum supported version of the Azure instance metadata API.
+	minimumSupportedAPIVersion = "2019-06-04"
+)
+
+// InstanceMetadataClient is a client for Azure instance metadata.
+type InstanceMetadataClient struct {
+	baseURL    string
+	apiVersion string
+}
+
+// InstanceMetadataClientOption allows setting options as functional arguments to an InstanceMetadataClient.
+type InstanceMetadataClientOption func(client *InstanceMetadataClient)
+
+// WithBaseURL sets the base URL for the metadata client. Used in tests.
+func WithBaseURL(url string) InstanceMetadataClientOption {
+	return func(client *InstanceMetadataClient) {
+		client.baseURL = url
+	}
+}
+
+// NewInstanceMetadataClient creates a new instance metadata client.
+func NewInstanceMetadataClient(opts ...InstanceMetadataClientOption) *InstanceMetadataClient {
+	client := &InstanceMetadataClient{}
+	for _, opt := range opts {
+		opt(client)
+	}
+
+	if client.baseURL == "" {
+		client.baseURL = imdsURL
+	}
+	return client
+}
+
+func (client *InstanceMetadataClient) GetType() types.InstanceMetadataType {
+	return types.InstanceMetadataTypeAzure
+}
+
+// selectVersion sets the Azure instance metadata API version.
+func (client *InstanceMetadataClient) selectVersion(ctx context.Context) error {
+	versions, err := client.getVersions(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// Use the most recent supported API version.
+	targetVersion, err := selectVersion(versions, minimumSupportedAPIVersion)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	client.apiVersion = targetVersion
+	return nil
+}
+
+// getRawMetadata gets the raw metadata from a specified path.
+func (client *InstanceMetadataClient) getRawMetadata(ctx context.Context, route string) ([]byte, error) {
+	httpClient := &http.Client{Transport: &http.Transport{Proxy: nil}}
```

Comment:
You can probably just use this https://github.com/gravitational/teleport/blob/4d25801fab927f3dc14b6eed75404da237969d66/lib/defaults/defaults.go#L776